lib/src/block_syntaxes/footnote_def_syntax.dart

```diff
@@ -0,0 +1,74 @@
+import '../ast.dart' show Element, Node;
+import '../block_parser.dart' show BlockParser;
+import '../line.dart';
+import '../patterns.dart' show footnotePattern;
+import 'block_syntax.dart' show BlockSyntax;
+import 'empty_block_syntax.dart';
+import 'paragraph_syntax.dart';
+import 'setext_header_syntax.dart';
+
+class FootnoteDefSyntax extends BlockSyntax {
+  const FootnoteDefSyntax();
+
+  @override
+  RegExp get pattern => footnotePattern;
+
+  @override
+  Node? parse(BlockParser parser) {
+    final current = parser.current.content;
+    final match = pattern.firstMatch(current)!;
+    final label = match[2]!;
+    final refs = parser.document.footnoteReferences;
+    refs[label] = 0;
+
+    final id = Uri.encodeComponent(label);
+    parser.advance();
+    final lines = [
+      Line(current.substring(match[0]!.length)),
+      ...parseChildLines(parser),
+    ];
+    // Linebreak in secondary paragraph did not match LineBreakSyntax
+    // so there is no `<br />`, unlike github.
+    final children = BlockParser(lines, parser.document).parseLines();
+    return Element('li', children)
+      ..attributes['id'] = 'fn-$id'
+      ..attributes['_label_'] = label;
+  }
+
+  @override
+  List<Line> parseChildLines(BlockParser parser) {
+    final children = <String>[];
+    var nextIsBlock = false;
+    while (!parser.isDone) {
+      final line = parser.current.content;
+      if (line.trim().isEmpty) {
+        children.add(line);
+        parser.advance();
+        nextIsBlock = true;
+        continue;
+      } else if (line.startsWith('    ')) {
+        children.add(line.substring(4));
+        parser.advance();
+        nextIsBlock = false;
+      } else if (nextIsBlock || _isBlock(parser, line)) {
+        break;
+      } else {
+        children.add(line);
+        parser.advance();
+      }
+    }
+    return children.map(Line.new).toList(growable: false);
+  }
+
+  static const _invalidSecondaryBlock = <Type>[
+    EmptyBlockSyntax,
+    ParagraphSyntax,
+    SetextHeaderSyntax,
+  ];
+
+  static bool _isBlock(BlockParser parser, String line) {
+    final syntaxList = parser.blockSyntaxes
+        .takeWhile((s) => !_invalidSecondaryBlock.contains(s.runtimeType));
```

Comment:
I don't want to use runtimeType here, as that would not account for (theoretical, to be sure) subtypes.

I think we can use `.whereType` and a late instance field to compute this once per parse. Something like:

```dart
late final secondaryBlockSyntaxes =
  [...parser.blockSyntaxes]
    .removeWhere((e) = _invalidSecondaryBlock.contains(e));
```

WDYT?

Reply:
> seems have to use runtimeType, we actually expect a iterable with same base type `BlockSyntax` but different subtype, `whereType` could only give us a iterable with same type.

Sorry I don't understand this. You might be misunderstanding what `whereType` does. `[1, 2.1, 'hello'].whereType<num>()` will yield both `1` and `2.1`.

> And for a late instance, I use `??=` to assign a local variable, a class field would make `FootnoteDefSyntax` none const constructor any more.

OK a local variable is ok here.

Reply:
I use BlockSytax's `pattern` field to exclude `EmptyBlockSyntax`, `ParagraphSyntax` and `SetextHeaderSyntax`, no `runtimeType` any more.